### PR TITLE
slicing operation for DVector/ByteArray/RawArray: Dvector.subarray()

### PR DESCRIPTION
**Pull request by gau-veldt**
This PR is related to #5701

Adds a slicing operation to DVector:
`DVector<T>& DVector<T>::subarray(int from,int to)`

Adds the two binds necessary to appear as `RawArray.subarray()` in GDScript

### CODE REVIEW

**Comment by gau-veldt** [comment]
Added handling for -1 for p_to as well the documentation for the method in classes.xml

**Comment by gau-veldt** [comment]
This is incredibly useful for networking code for chopping of parts of buuffers successfully sent by the nonbloocking 'partial' xfer methods of `StreamPeer`'s

File: core/dvector.h [hunk 1]
```diff
@@ -264,6 +264,9 @@ class DVector {
 
 	DVector<T> subarray(int p_from, int p_to) {
 
+		if (p_to==-1 && size()>0) {
```
Comment (akien-mga) [inline core/dvector.h]: Can't you use #4861 instead? Because if we're supporting -1, then we should support -2, -3, etc.

**Comment by gau-veldt** [comment]
Okay I've amended the code to handle <0 to mean size()-index for both indices.  Should I add .kitchen_sink() while I'm at it? j/k ;)

**Comment by gau-veldt** [comment]
left-deletes can be done in script with the following hack though it copies the data three times in doing so (why I call it a hack).

The proposed slicing addition only copies the data once.  As far as I can see and since scripts don't work at the low level Write/Read lock level the absence of such a method and therefore requiring a script level loop using .set() or .push_back() to implement .subarray() is really bad and results in the ByteArray copy-on-writing each pass through the script's loop 1+2+3+4+5+6+7+8+9+..+N. O( (n*(n+1))/2 ) == O ( n^2 ).  For 10 KB of data (barely one sprite frame worth of PNG sent over network) that's 50 MB (50 or so minutes of .ogg) of wasted copying!!!  Consider that when evaluating the benefit of the DVector.subarray() addition that gets the job done with only one copy ever made O(N) of the original data.

    func left_delete(raw, amount):
        if amount<1:
            return raw
        if amount>=raw.size():
            raw.resize(0)
            return raw
        raw.invert()
        raw.resize(raw.size()-amount)
        raw.invert()
        return raw

**Comment by gau-veldt** [comment]
moved to #5879

**Pull request closed without merging**